src/shogun/lib/any.h

```diff
@@ -327,18 +327,40 @@ namespace shogun
 		template <class T, class S>
 		inline auto free_array(T* ptr, S size)
 		{
+			if (!ptr)
+			{
+				return;
+			}
 			SG_FREE(ptr);
 		}
 
 		template <class S>
 		inline auto free_array(CSGObject** ptr, S size)
 		{
+			if (!ptr)
+			{
+				return;
+			}
 			for (S i = 0; i < size; ++i)
 			{
 				ptr[i]->unref();
 			}
 			SG_FREE(ptr);
 		}
+
+		template <class T>
+		inline auto copy_array(T* begin, T* end, T* dst)
+		{
+			std::copy(begin, end, dst);
+		}
+
+		template <class T>
+		inline auto copy_array(T** begin, T** end, T** dst)
+		    -> decltype((void)begin[0]->clone())
+		{
+			std::transform(
+			    begin, end, dst, [](auto each) { return each->clone(); });
```

Comment:
What if some elements of the array are null?

Reply:
Good catch